src/dns.rs

```diff
@@ -0,0 +1,305 @@
+use core::mem;
```

Comment:
I must say, I *LOVE* the idea of having a Captive Portal!
What I don't love that much:
* Overall, the code feels very low level (as if we are coding in C, via Rust) - I'll comment more below
* It is part of `esp-idf-svc` but I'm not sure what it is doing here? This is a generic DNS server, which can - just as well - be coded against Rust's STD APIs (correct me if I'm wrong!) - or better yet - against the [embedded-nal ](https://github.com/rust-embedded-community/embedded-nal/blob/master/src/stack/udp.rs#L50) so that it does have a `no_std` story as well. (`embedded-nal` traits do have STD implementations too, and if these are lacking, we can implement / patch / upstream our own changes.). So I'm very much in favor of moving this into `embedded-svc` (which depends on `no-std-net` already anyway) - possibly behind a feature toggle.
* As for what would be the use-case of that? Well, I might want to use a captive dns portal on the Raspberry Pi too?

Reply:
The code is indeed C converted to Rust, but not quite at the point where I'm super happy with the "Rustyness" of it yet

Sure, it would be nice to have it generic.
But my use case wasn't quite as generic, so I implemented it specifically for the esp right now.

`embedded-nal` seems nice, hadn't heard of it before.

I feel like someone else must have already implemented a captive portal dns in rust that's more generic, though. (not that I checked and not that I'm opposed to make this more generic)